bindings/rust/s2n-tls/src/callbacks/async_cb.rs

```diff
@@ -0,0 +1,154 @@
+    }
+}
+
+pin_project! {
+    /// A wrapper around an optional [`ConnectionFuture`]
+    /// which either polls the future or immediately reports success.
+    struct OptionalFuture{
+        option: Option<Pin<Box<dyn ConnectionFuture>>>,
+    }
+}
+
+impl OptionalFuture {
+    fn new(input: ConnectionFutureResult) -> Self {
+        match input {
+            Ok(option) => OptionalFuture { option },
+            Err(error) => {
+                let error = Some(error);
+                OptionalFuture {
+                    option: Some(Box::pin(ErrorFuture { error })),
+                }
+            }
+        }
+    }
+}
+
+impl ConnectionFuture for OptionalFuture {
+    fn poll(
+        mut self: Pin<&mut Self>,
+        conn: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>> {
+        match self.option.as_mut() {
+            Some(future) => future.as_mut().poll(conn, ctx),
+            None => Poll::Ready(Ok(())),
+        }
+    }
+}
+
+/// Any work necessary after the callback completes.
+//
+// We do not expect any callback except [`ClientHelloCallback`] to require MarkDone.
+// More recent callbacks follow a different model that doesn't require separate cleanup.
+//
+// This enum is sufficient while only ClientHello is special-cased, but will not
+// scale well. If we need more MarkDone variants, then we should consider a different
+// solution, like another stored future.
+#[non_exhaustive]
+#[derive(PartialEq)]
+enum MarkDone {
+    ClientHello,
+    // None,
+}
+
+pin_project! {
+    // Stores the [`ConnectionFuture`] and associated state.
+    pub(crate) struct AsyncCallback {
+        #[pin]
+        future: OptionalFuture,
+        cleanup: MarkDone,
+    }
+}
+
+impl AsyncCallback {
+    pub(crate) fn poll(
+        self: Pin<&mut Self>,
+        conn: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>> {
+        let this = self.project();
+        let poll = this.future.poll(conn, ctx);
+        if let Poll::Ready(Ok(())) = poll {
+            if this.cleanup == &MarkDone::ClientHello {
+                conn.mark_client_hello_cb_done()?;
+            }
+        }
+        poll
+    }
+
+    pub(crate) fn trigger_client_hello_cb(
+        future: ConnectionFutureResult,
+        conn: &mut Connection,
+    ) -> CallbackResult {
+        let future = OptionalFuture::new(future);
+        let cleanup = MarkDone::ClientHello;
+        let callback = AsyncCallback { future, cleanup };
+        conn.set_async_callback(callback);
+        CallbackResult::Success
+    }
+
+    // pub(crate) fn trigger(
+    //    future: ConnectionFutureResult,
+    //    conn: &mut Connection,
+    //) -> CallbackResult {
+    //    let future = OptionalFuture::new(future);
+    //    let cleanup = MarkDone::None;
+    //    let callback = AsyncCallback { future, cleanup };
+    //    conn.set_async_callback(callback);
+    //    CallbackResult::Success
+    //}
```

Comment:
I'm assuming this is going to be used for a later PR? I would say to probably include it there, if so

Reply:
I just saw that in your call outs. Not a big deal, then.

Reply:
Yeah I usually don't liked commented out code, but I'm un-commenting it in literally the next PR and without it it looks like this is still pretty client-hello specific ;) I just didn't want everyone calling trigger to have to pass in the MarkDone, even though only the ClientHello needs it.